File: packages/live-update/android/src/main/java/io/capawesome/capacitorjs/plugins/liveupdate/LiveUpdatePreferences.java
package io.capawesome.capacitorjs.plugins.liveupdate;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.UUID;

public class LiveUpdatePreferences {

    @NonNull
    private final Context context;

    @NonNull
    private final SharedPreferences.Editor settingsEditor;

    private final String blockedBundleIdsKey = "blockedBundleIds"; // DO NOT CHANGE
    private final String channelKey = "channel"; // DO NOT CHANGE
    private final String deviceIdKey = "deviceId"; // DO NOT CHANGE
    private final String customIdKey = "customId"; // DO NOT CHANGE
    private final String previousBundleIdKey = "previousBundleId"; // DO NOT CHANGE

    public LiveUpdatePreferences(@NonNull Context context) {
        this.context = context;
        this.settingsEditor = context.getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE).edit();
    }

    @Nullable
    public String getChannel() {
        return context.getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE).getString(channelKey, null);
    }

    @Nullable
    public String getCustomId() {
        return context.getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE).getString(customIdKey, null);
    }

    @Nullable
    public String getDeviceIdForApp(@Nullable String appId) {
        if (appId == null) {
            return context
                .getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE)
                .getString(deviceIdKey, null);
        } else {
            return context
                .getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE)
                .getString(deviceIdKey + "_" + appId, null);
        }
    }

    @Nullable
    public String getPreviousBundleId() {
        return context
            .getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE)
            .getString(previousBundleIdKey, null);
    }

    @Nullable
    public String getBlockedBundleIds() {
        return context
            .getSharedPreferences(LiveUpdatePlugin.SHARED_PREFERENCES_NAME, Activity.MODE_PRIVATE)
            .getString(blockedBundleIdsKey, null);
    }

    public void setBlockedBundleIds(@Nullable String blockedBundleIds) {
        if (blockedBundleIds == null) {
            settingsEditor.remove(blockedBundleIdsKey);
        } else {
            settingsEditor.putString(blockedBundleIdsKey, blockedBundleIds);
        }
        settingsEditor.apply();
    }

    public void setChannel(@Nullable String channel) {
        if (channel == null) {
            settingsEditor.remove(channelKey);
        } else {
            settingsEditor.putString(channelKey, channel);
        }
        settingsEditor.apply();
    }

    public void setCustomId(@Nullable String customId) {
        if (customId == null) {
            settingsEditor.remove(customIdKey);
        } else {
            settingsEditor.putString(customIdKey, customId);
        }
        settingsEditor.apply();
    }

    public void setDeviceIdForApp(@Nullable String appId, @NonNull String deviceId) {
        if (appId == null) {
            settingsEditor.putString(deviceIdKey, deviceId);
        } else {
            settingsEditor.putString(deviceIdKey + "_" + appId, deviceId);
        }
        settingsEditor.apply();
    }

    public void setPreviousBundleId(@Nullable String bundleId) {
        if (bundleId == null) {
            settingsEditor.remove(previousBundleIdKey);
        } else {
            settingsEditor.putString(previousBundleIdKey, bundleId);
        }
        settingsEditor.apply();
    }
}

File: packages/live-update/ios/Plugin/LiveUpdatePreferences.swift
import Foundation

public class LiveUpdatePreferences: NSObject {
    private static let userDefaultsPrefix = "CapawesomeLiveUpdate" // DO NOT CHANGE

    private let appIdKey = "appId" // DO NOT CHANGE
    private let blockedBundleIdsKey = "blockedBundleIds" // DO NOT CHANGE
    private let channelKey = "channel" // DO NOT CHANGE
    private let customIdKey = "customId" // DO NOT CHANGE
    private let lastVersionCodeKey = "lastVersionCode" // DO NOT CHANGE
    private let lastVersionNameKey = "lastVersionName" // DO NOT CHANGE
    private let previousBundleIdKey = "previousBundleIdKey" // DO NOT CHANGE

    public func getAppId() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: appIdKey))
    }

    public func getBlockedBundleIds() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: blockedBundleIdsKey))
    }

    public func getChannel() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: channelKey))
    }

    public func getCustomId() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: customIdKey))
    }

    public func getLastVersionCode() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: lastVersionCodeKey))
    }

    public func getLastVersionName() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: lastVersionNameKey))
    }

    public func getPreviousBundleId() -> String? {
        return UserDefaults.standard.string(forKey: applyPrefix(to: previousBundleIdKey))
    }

    public func setAppId(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: appIdKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: appIdKey))
        }
        UserDefaults.standard.synchronize()
    }

    public func setBlockedBundleIds(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: blockedBundleIdsKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: blockedBundleIdsKey))
        }
        UserDefaults.standard.synchronize()
    }

    public func setChannel(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: channelKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: channelKey))
        }
        UserDefaults.standard.synchronize()
    }

    public func setCustomId(_ value: String) {
        UserDefaults.standard.set(value, forKey: applyPrefix(to: customIdKey))
        UserDefaults.standard.synchronize()
    }

    public func setLastVersionCode(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: lastVersionCodeKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: lastVersionCodeKey))
        }
        UserDefaults.standard.synchronize()
    }

    public func setLastVersionName(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: lastVersionNameKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: lastVersionNameKey))
        }
        UserDefaults.standard.synchronize()
    }

    public func setPreviousBundleId(_ value: String?) {
        if let value = value {
            UserDefaults.standard.set(value, forKey: applyPrefix(to: previousBundleIdKey))
        } else {
            UserDefaults.standard.removeObject(forKey: applyPrefix(to: previousBundleIdKey))
        }
        UserDefaults.standard.synchronize()
    }

    private func applyPrefix(to key: String) -> String {
        return LiveUpdatePreferences.userDefaultsPrefix + "." + key
    }
}
